File: leetcode/FactorialTrailingZeroes.cpp
// Works. But Solution2 is the best.
class Solution3 {
public:
    int trailingZeroes(int n) {
        int ct = 0;
        for (int i = 5; i <= n; ++ i) {
            int k = i;
            while (k % 5 == 0 && k > 0) {
                ++ ct;
                k /= 5;
            }
        }
        return ct;
    }
};

// Works too. Saved one division than Solution.
class Solution2 {
public:
    int trailingZeroes(int n) {
        int ct = 0;
        while (n > 0) {
            n /= 5;
            ct += n;
        }
        return ct;
    }
};

//
// This works.
//
class Solution {
public:
    int trailingZeroes(int n) {
        int v = 0;
        
        while (n > 0) {
            v += n/5;
            n = n/5;
        }
        
        return v;
    }
};

/**
Factorial Trailing Zeroes 
Given an integer n, return the number of trailing zeroes in n!.

Note: Your solution should be in logarithmic time complexity.
 */

File: leetcode_java/FactorialTrailingZeroes.java
/**
 * x = 1 + Math.floor(n/5) + Math.floor(n/5^2) + ...
 */
public class Solution {
    public int trailingZeroes(int n) {
        int x = 0;
        while (n > 0) {
            x += n/5;
            n /= 5;
        }
        
        return x;
    }
}

/**
Factorial Trailing Zeroes
Difficulty: Easy

Given an integer n, return the number of trailing zeroes in n!.

Note: Your solution should be in logarithmic time complexity.
 */
